pkg/controller/synchronize.go

```diff
@@ -0,0 +1,93 @@
+package controller
+
+import (
+	"fmt"
+	"github.com/oslokommune/okctl/pkg/controller/reconsiler"
+	"github.com/oslokommune/okctl/pkg/controller/resourcetree"
+	"github.com/spf13/afero"
+)
+
+type SynchronizeOpts struct {
+	DesiredTree *resourcetree.ResourceNode
+	
+	ReconsiliationManager *reconsiler.ReconsilerManager
+	
+	Fs *afero.Afero
+	OutputDir string
+
+	GithubGetter reconsiler.GithubGetter
+	GithubSetter reconsiler.GithubSetter
+
+	CIDRGetter StringFetcher
+	PrimaryHostedZoneGetter HostedZoneFetcher
+}
+
+// Synchronize knows how to discover differences between desired and actual state and rectify them
+func Synchronize(opts *SynchronizeOpts) error {
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeCluster, CreateClusterStateRefresher(
+		opts.Fs,
+		opts.OutputDir,
+		opts.CIDRGetter,
+	))
+	
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeALBIngress, CreateALBIngressControllerRefresher(
+		opts.Fs,
+		opts.OutputDir,
+	))
+
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeExternalDNS, CreateExternalDNSStateRefresher(
+		opts.PrimaryHostedZoneGetter,
+	))
+	
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeIdentityManager, CreateIdentityManagerRefresher(
+		opts.PrimaryHostedZoneGetter,
+	))
+
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeGithub, CreateGithubStateRefresher(
+		opts.GithubGetter,
+		opts.GithubSetter,
+	))
+	
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeArgoCD, CreateArgocdStateRefresher(
+		opts.PrimaryHostedZoneGetter,
+	))
+
+	currentStateGraphOpts, err := NewCreateCurrentStateGraphOpts(opts.Fs, opts.OutputDir, opts.GithubGetter)
+	if err != nil {
+	    return fmt.Errorf("unable to get existing services: %w", err)
+	}
+
+	currentStateTree := CreateCurrentStateGraph(currentStateGraphOpts)
+
+	diffGraph := *opts.DesiredTree
+
+	diffGraph.ApplyFunction(applyCurrentState, currentStateTree)
```

Comment:
I just had to try to see how this would look without passing a function:

```go
func applyStateOnTree(desired *resourcetree.ResourceNode, current *resourcetree.ResourceNode) {
	for _, desiredChild := range desired.Children {
		applyStateOnTree(desiredChild, current)
	}

	targetNode := current.GetNode(desired)
	if desired.State == targetNode.State {
		desired.State = resourcetree.ResourceNodeStateNoop
	}
}
```

(Btw this code could be buggy, haven't tested it.)


This code more compact (as you don't need the diffGraph.ApplyFunction method anymore). But you lose the current tree iteration ability - getting an equivalent node from the target tree. I'm just wondering if you have considered simpler, but less future-proof alternatives - no need to change anything.

Reply:
Yeah. I actually rewrote a function like that for the more future proof one.

I know that at some point soon I must implement delete cluster, in which I'll create a applyDeletedNodesInformation function and use the same "iterator"-function with. I hope that makes sense